Optimization and debug

cifren opened the pull request:
I have added several modifications in order to improve debug and speed.

I have as well renamed cas.properties in order to be able to avoid to send cas.properties into github accidently like several projects does

And I have moved logs into a log folder to make it less messy when executing several times the web server

cifren commented:
I made it as well easier to use the first time someone comes to that project, default usable configuration on first start

mmoayyed commented:
We cannot accept the default configuration as you have it, sadly. Deployers are expected to generate everything on their own locally, specially the keys. We have run into severe sec vulns before where folks simply ignored that step and ran with a bunch of default sample keys. 

Remove the dist file please and all related changed.

mmoayyed commented:
That said, there are grand plans in place to automate all of this! Stay tuned for more info, or tune into open apereo 2017 to learn more

mmoayyed commented on .gitignore:
```diff
@@ -19,3 +19,6 @@ build
 MANIFEST.MF
+
+.c9
+cas.properties
```
this needs to be removed

cifren replied:
Yes I thought that part was too much

mmoayyed commented on README.md:
```diff
@@ -32,6 +32,13 @@ Study material:
 - https://docs.gradle.org/current/userguide/artifact_dependencies_tutorial.html
 - https://docs.gradle.org/current/userguide/dependency_management.html
 
+## First instalaltion
```
all should be removed

mmoayyed commented on README.md:
````diff
@@ -51,6 +58,8 @@ If you are on a `SNAPSHOT` version, you can force redownloads of modules/depende
  ./gradlew[.bat] clean build --parallel --refresh-dependencies
 ```
 
+Some options are in 'gradle.properties', and you add your own in order to improve gradle speed
````
you need to be a bit more explicit. what options? and what is expected to be modified? are those really things that cas should document or a gradle-person should already know?

cifren replied:
The limit between what should know gradle user and what cas should provide is not easy to define. I added this comment in order to make the user aware that the option "--parallel" has been added to the gradle.properties. 

The documentation was talking about it, I thought it was relevant to add some comment for that.

mmoayyed replied:
I suppose: if you decide to use the gradle cas overlay then you must be familiar with gradle to some extent. we don't need to teach you as the deployer what maven or gradle those. Best we can do is to refer folks to their official documentation and have them learn it. I think adding a link simply is good, because if and when gradle changes X on their side, we shouldn't carry the maintenance burden to update the same duplicated thing in our docs.

mmoayyed commented on README.md:
````diff
@@ -120,3 +130,32 @@ Be careful with this method of deployment. `bootRun` is not designed to work wit
+
+```bash
+# shortcut for clean, build and explodeWar (mainly for debug purposes)
+./gradlew[.bat] buildit
````
it's probably best if all of this goes into a bash script of some kind like the Maven overlay

cifren replied:
Gradle is not there to replace the bash from maven ? That is like a shorcut, I mainly use it in my IDE. It could be removed if not needed

mmoayyed replied:
The bash script acts as a bridge to simplify the gradle commands; it is not a replacement. It is so that you can run "build.sh debug". It is the shortcut, and is to keep things possibly more consistent between the projects. Bash arguably is easier for folks to deal with, as "everyone knows how to run bash script" and can decipher it, rather than running "gradlew builtit" or anything similar.

mmoayyed commented on cas/build.gradle:
```diff
@@ -74,11 +83,63 @@ task generateKeys {
+            args = [
+                "build/libs/cas.war",
+                "--logging.config=file:///home/cas/etc/cas/config/log4j2.xml",
+                "--spring.cloud.config.server.native.searchLocations=file:///home/cas/etc/cas/config"
```
this is also not needed as it's the default

cifren replied:
I don't understand why "/etc/cas/config" is default, would that not be easier to keep everything in the same folder where you do not have to have sudo permission on the server ? So that means to avoid the copyConfig and maybe use relative path to a folder "/cas/config" ?

cifren replied:
As well this config, with the copy, forces the user to run a build anytime it has to change the cas.properties, which "--spring.cloud.config.server.native.searchLocations=" avoids

mmoayyed commented on cas/build.gradle:
```diff
@@ -74,11 +83,63 @@ task generateKeys {
+                ]
+            args = [
+                "build/libs/cas.war",
+                "--logging.config=file:///home/cas/etc/cas/config/log4j2.xml",
```
same thing here. defaults

mmoayyed commented on etc/cas/config/cas.properties:
```diff
@@ -1,7 +0,0 @@
-cas.adminPagesSecurity.ip=127\.0\.0\.1
-
-logging.config: file:/etc/cas/config/log4j2.xml
-# cas.serviceRegistry.config.location: classpath:/services
```
this will also need to be restored back

cifren commented:
I understand for the keys, it is not something that could block the user anyway.

In that pull request I tried to make the project works on first try, which I didn't success when I started. 

Usually the cas web server crash systematically on first boot, mainly because the /etc/cas/keytab is not existing, other execption like the https is possible to be accessed. It doesn't make it accessible at first.

You are right I will remove all superficial changes.

cifren commented:
I have modified both commits, you can review again, I made it lighter.

I have removed a commit, some comments disappeared.

I don't understand why do you choose to use '/etc/cas/config" instead of the same folder (like "casProject/cas/config") and make that copy an option. It adds a another copy task that needs to be executed each time the cas.properties is modified and the web server needs to be restarted, which doesn't make it faster. And as well sudo is neccessary to make the copy... or at least to prepare the folder "/etc/cas" with the correct permissions.

mmoayyed commented:
Thank you. I'll review again.

The idea usually is to pick something and stick with it. "Why not casProject/cas/config" is immediately met with "why not cas-workspace/cas/config" and then met with "why not /etc/cas/config or etc/cas/config-whatever". The goal as it find a place as the convention and stick with it. You will always find folks who don't like that convention, and they are allowed to make the path be different.

But your question is one of workflow: typically the overlay directory is managed in source control, so you can keep your configuration as well as the build script inside git/svn and carry on. Encrypt sensitive settings, etc. You want to make all changes in the "workspace" directory and run things from there, and have it be automated. You clone the repo, run the build and it knows what to do. You shouldn't have to learn to modify "etc/cas/config" or anything, and you shouldn't have to teach your colleagues or document it even to a degree. 

The build does it all itself. 

In most cases, the cas.properties file changes do not require a restart...and if you use embedded, the process is all contained within a single command: "sync the configuration, build the project, start the server, and do it again". The goal is to reduce manual work and automate. 

If you find yourself more comfortable with CDing into /etc/cas/config directory, making changes and restarting tomcat for instance manually outside the build script, (and occasionally building the script when needed), and then find a way to keep your settings in source control somehow, along with the core overlay project, and then document it's all set up...that's all great. Please do. That however would not be my choice. 

HTH.

cifren commented:
Done, "first installation" removed and "buildit" too

mmoayyed commented:
Thank you. Great work!
